File: agent/internal/restic/restic.go
package restic

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

// Wrapper provides methods to interact with restic.
type Wrapper struct {
	Repository string
	Password   string
}

// New creates a new restic wrapper.
func New(repository, password string) *Wrapper {
	return &Wrapper{
		Repository: repository,
		Password:   password,
	}
}

// Snapshot represents a restic snapshot.
type Snapshot struct {
	ID       string   `json:"id"`
	ShortID  string   `json:"short_id"`
	Time     string   `json:"time"`
	Tree     string   `json:"tree"`
	Paths    []string `json:"paths"`
	Hostname string   `json:"hostname"`
	Tags     []string `json:"tags"`
}

// BackupResult holds the output of a backup operation.
type BackupResult struct {
	FilesNew    int    `json:"files_new"`
	FilesChanged int   `json:"files_changed"`
	FilesTotal  int    `json:"files_total"`
	BytesAdded  int64  `json:"bytes_added"`
	BytesTotal  int64  `json:"bytes_total"`
	SnapshotID  string `json:"snapshot_id"`
}

// Init initializes a new restic repository.
func (r *Wrapper) Init() error {
	_, err := r.run("init", "----repo-version", "2")
	if err != nil && !strings.Contains(err.Error(), "already initialized") {
		return fmt.Errorf("restic init: %w", err)
	}
	return nil
}

// Backup runs a backup of the given paths.
func (r *Wrapper) Backup(paths []string, tags []string) (*BackupResult, error) {
	args := []string{"backup"}
	args = append(args, paths...)
	for _, tag := range tags {
		args = append(args, "--tag", tag)
	}
	args = append(args, "--json")

	output, err := r.run(args...)
	if err != nil {
		return nil, fmt.Errorf("restic backup: %w", err)
	}

	// Parse the JSON output (last line is the summary)
	lines := strings.Split(strings.TrimSpace(output), "\n")
	for i := len(lines) - 1; i >= 0; i-- {
		var msg struct {
			MessageType string `json:"message_type"`
			FilesNew    int    `json:"files_new"`
			FilesChanged int   `json:"files_changed"`
			FilesTotal  int    `json:"total_files"`
			BytesAdded  int64  `json:"bytes_added"`
			BytesTotal  int64  `json:"total_bytes"`
			SnapshotID  string `json:"snapshot_id"`
		}
		if json.Unmarshal([]byte(lines[i]), &msg) == nil && msg.MessageType == "summary" {
			return &BackupResult{
				FilesNew:    msg.FilesNew,
				FilesChanged: msg.FilesChanged,
				FilesTotal:  msg.FilesTotal,
				BytesAdded:  msg.BytesAdded,
				BytesTotal:  msg.BytesTotal,
				SnapshotID:  msg.SnapshotID,
			}, nil
		}
	}

	return &BackupResult{}, nil
}

// Snapshots lists all snapshots in the repository.
func (r *Wrapper) Snapshots() ([]Snapshot, error) {
	output, err := r.run("snapshots", "--json")
	if err != nil {
		return nil, fmt.Errorf("restic snapshots: %w", err)
	}

	var snapshots []Snapshot
	if err := json.Unmarshal([]byte(output), &snapshots); err != nil {
		return nil, fmt.Errorf("parsing snapshots: %w", err)
	}
	return snapshots, nil
}

// Restore restores files from a snapshot.
func (r *Wrapper) Restore(snapshotID, target, includePath string) error {
	args := []string{"restore", snapshotID, "--target", target}
	if includePath != "" {
		args = append(args, "--include", includePath)
	}
	_, err := r.run(args...)
	if err != nil {
		return fmt.Errorf("restic restore: %w", err)
	}
	return nil
}

// List lists files in a snapshot.
func (r *Wrapper) List(snapshotID, path string) ([]string, error) {
	args := []string{"ls", snapshotID}
	if path != "" {
		args = append(args, path)
	}
	args = append(args, "--json")

	output, err := r.run(args...)
	if err != nil {
		return nil, fmt.Errorf("restic ls: %w", err)
	}

	var files []string
	for _, line := range strings.Split(output, "\n") {
		var node struct {
			MessageType string `json:"message_type"`
			Name        string `json:"name"`
			Path        string `json:"path"`
			NodeType    string `json:"node_type"`
		}
		if json.Unmarshal([]byte(line), &node) == nil && node.MessageType == "node" {
			files = append(files, node.Path)
		}
	}
	return files, nil
}

// Forget removes snapshots according to retention policy.
func (r *Wrapper) Forget(keepDaily, keepWeekly, keepMonthly, keepYearly int) error {
	args := []string{"forget"}
	if keepDaily > 0 {
		args = append(args, "--keep-daily", fmt.Sprintf("%d", keepDaily))
	}
	if keepWeekly > 0 {
		args = append(args, "--keep-weekly", fmt.Sprintf("%d", keepWeekly))
	}
	if keepMonthly > 0 {
		args = append(args, "--keep-monthly", fmt.Sprintf("%d", keepMonthly))
	}
	if keepYearly > 0 {
		args = append(args, "--keep-yearly", fmt.Sprintf("%d", keepYearly))
	}
	args = append(args, "--prune")

	_, err := r.run(args...)
	if err != nil {
		return fmt.Errorf("restic forget: %w", err)
	}
	return nil
}

// Stats returns repository statistics.
func (r *Wrapper) Stats() (map[string]interface{}, error) {
	output, err := r.run("stats", "--json")
	if err != nil {
		return nil, fmt.Errorf("restic stats: %w", err)
	}

	var stats map[string]interface{}
	if err := json.Unmarshal([]byte(output), &stats); err != nil {
		return nil, fmt.Errorf("parsing stats: %w", err)
	}
	return stats, nil
}

// Check verifies repository integrity.
func (r *Wrapper) Check() error {
	_, err := r.run("check")
	return err
}

// run executes a restic command with the configured repository and password.
func (r *Wrapper) run(args ...string) (string, error) {
	cmdArgs := append([]string{"-r", r.Repository}, args...)

	cmd := exec.Command("restic", cmdArgs...)
	cmd.Env = append(os.Environ(), "RESTIC_PASSWORD="+r.Password)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		if stderr.Len() > 0 {
			return "", fmt.Errorf("%s: %s", err, strings.TrimSpace(stderr.String()))
		}
		return "", err
	}
	return stdout.String(), nil
}

File: server/src/lib/restic.ts
/**
 * Restic command wrapper — server-side operations.
 * Used for restore operations and repository management.
 */

import { $ } from "bun"

export interface ResticSnapshot {
  id: string
  short_id: string
  time: string
  tree: string
  paths: string[]
  hostname: string
  tags: string[]
}

export interface ResticStats {
  total_size: number
  total_file_count: number
  snapshots_count: number
}

export interface ResticFileNode {
  name: string
  path: string
  type: string
  mode: string
  size: number
  time: string
}

/**
 * Run a restic command with the given repository and password.
 */
async function runRestic(repo: string, password: string, args: string[]): Promise<string> {
  try {
    const result = await $`restic -r ${repo} ${args}`.env({
      ...process.env,
      RESTIC_PASSWORD: password,
    }).text()
    return result
  } catch (err: any) {
    throw new Error(`restic error: ${err.stderr || err.message}`)
  }
}

/**
 * Initialize a restic repository.
 */
export async function initRepo(repo: string, password: string): Promise<void> {
  try {
    await runRestic(repo, password, ["init", "--repo-version", "2"])
  } catch (err: any) {
    if (!err.message?.includes("already initialized")) {
      throw err
    }
  }
}

/**
 * List all snapshots in a repository.
 */
export async function listSnapshots(repo: string, password: string): Promise<ResticSnapshot[]> {
  const output = await runRestic(repo, password, ["snapshots", "--json"])
  return JSON.parse(output)
}

/**
 * List files in a snapshot.
 */
export async function listFiles(
  repo: string,
  password: string,
  snapshotId: string,
  path: string = "/"
): Promise<ResticFileNode[]> {
  const output = await runRestic(repo, password, ["ls", snapshotId, path, "--json"])
  const nodes: ResticFileNode[] = []
  for (const line of output.split("\n")) {
    try {
      const msg = JSON.parse(line)
      if (msg.message_type === "node") {
        nodes.push({
          name: msg.name,
          path: msg.path,
          type: msg.node_type,
          mode: msg.mode || "",
          size: msg.size || 0,
          time: msg.time || "",
        })
      }
    } catch {}
  }
  return nodes
}

/**
 * Restore files from a snapshot.
 */
export async function restore(
  repo: string,
  password: string,
  snapshotId: string,
  target: string,
  includePath?: string
): Promise<void> {
  const args = ["restore", snapshotId, "--target", target]
  if (includePath) {
    args.push("--include", includePath)
  }
  await runRestic(repo, password, args)
}

/**
 * Get repository statistics.
 */
export async function getStats(repo: string, password: string): Promise<ResticStats> {
  const output = await runRestic(repo, password, ["stats", "--json"])
  return JSON.parse(output)
}

/**
 * Forget (prune) snapshots according to retention policy.
 */
export async function forget(
  repo: string,
  password: string,
  keepDaily: number = 7,
  keepWeekly: number = 4,
  keepMonthly: number = 6,
  keepYearly: number = 2
): Promise<void> {
  await runRestic(repo, password, [
    "forget",
    "--keep-daily", String(keepDaily),
    "--keep-weekly", String(keepWeekly),
    "--keep-monthly", String(keepMonthly),
    "--keep-yearly", String(keepYearly),
    "--prune",
  ])
}

/**
 * Check repository integrity.
 */
export async function check(repo: string, password: string): Promise<void> {
  await runRestic(repo, password, ["check"])
}

/**
 * Get the repository path for a job.
 */
export function getRepoPath(dataDir: string, jobId: string): string {
  return `${dataDir}/repos/${jobId}`
}
